crates/bls-crypto/src/bls/batch.rs

```diff
@@ -0,0 +1,86 @@
+use super::{PublicKey, Signature};
+
+// use algebra::{bls12_377::G1Projective, Field, PrimeField, ToBytes};
+use ark_bls12_377::{Fr, G1Projective};
+use ark_ff::{Field, PrimeField, ToBytes};
+
+use blake2s_simd::{Params, State};
+
+use crate::{BLSError, HashToCurve};
+
+#[derive(Default)]
+pub struct Batch {
+    entries: Vec<(PublicKey, Signature)>,
+    message: Vec<u8>,
+    extra_data: Vec<u8>,
+    size: usize,
+}
+
+impl Batch {
+    /// Constructs a new strict batch verifier context for a given message.
+    pub fn new(message: &[u8], extra_data: &[u8]) -> Batch {
+        Batch {
+            entries: Vec::<(PublicKey, Signature)>::new(),
+            message: message.to_vec(),
+            extra_data: extra_data.to_vec(),
+            size: 0,
+        }
+    }
+
+    pub fn add(&mut self, public_key: &PublicKey, signature: &Signature) {
+        self.entries
+            .push((public_key.to_owned(), signature.to_owned()));
+        self.size += 1
+    }
+
+    pub fn verify<H: HashToCurve<Output = G1Projective>>(
+        &self,
+        hash_to_g1: &H,
+    ) -> Result<(), BLSError> {
+        let mut public_keys = Vec::<&PublicKey>::new();
+        let mut signatures = Vec::<&Signature>::new();
+
+        let mut hash_states = Vec::<State>::new();
+        let mut hash_inputs = Vec::<Vec<u8>>::new();
+
+        self.entries.iter().for_each(|(pk, sig)| {
+            public_keys.push(&pk);
+            signatures.push(&sig);
+
+            hash_states.push(
+                Params::new()
+                    .personal(b"bvblssig")
+                    .hash_length(32)
+                    .to_state(),
+            );
+
+            // r <- H(pk || m || ad || sig)
+            let mut input = vec![];
+            pk.as_ref().write(&mut input).unwrap();
+            self.message.write(&mut input).unwrap();
+            self.extra_data.write(&mut input).unwrap();
+            sig.as_ref().write(&mut input).unwrap();
+            hash_inputs.push(input);
+        });
+
+        let security_bound = (128 + ((self.size as f64).log2().ceil() as usize) + 7) / 8; // in bytes
```

Comment:
Do we have some more generic functions for doing this? It might look cleaner and then we also don't have to check each time we convert bits to bytes and such.